File: 프로그래머스/lv2/68645. 삼각 달팽이/삼각 달팽이.java
class Solution {   
    public int[] solution(int n) {
        
        //하-우-좌상
        int dx[] = {0,1,-1}, dy[] = {1,0,-1};
        
        //피라미드 칸의 개수
        int numP = (n * (n + 1))/2;
        int[] answer = new int[numP];
        int[][] pyramid = new int[n][n];
        
        //시작칸, 순서
        int x = 0, y = 0, order = 1, d = 0;
        
        //남은 칸이 있다면
        while(numP > 0){
            pyramid[y][x] = order++;
            int nextX = x + dx[d];
            int nextY = y + dy[d];
            numP--;
            if( nextX >= n || nextX < 0 || nextY >= n || nextY < 0 || pyramid[nextY][nextX] != 0){
                d = (d + 1) % 3;
            }
            x = x + dx[d];
            y = y + dy[d];
        }
        
        //1차원 배열에 순서 저장
        int idx = 0;
        for(int i = 0; i < n; i++){
            for(int j = 0; j <= i; j++){
                answer[idx++] = pyramid[i][j];
            }
        }
        return answer;
    }
}

File: 프로그래머스/lv2/68645. 삼각 달팽이/삼각 달팽이.py
def solution(n):
    # 총 몇 칸인지 계산
    total = (n * (n + 1))/2
    answer = []
    
    # 2차원 리스트 생성
    triangle = [[0]*n for _ in range(n)]
    
    # 방향 저장하고 (하,우,좌상)
    dx, dy =  [0, 1, -1], [1, 0, -1]
    x, y, d, order = 0, 0, 0, 1
    
    # 방문할 칸이 남아있다면 탐색
    while total > 0 :
        triangle[y][x] = order
        order += 1
        total -= 1
        nx = x + dx[d]
        ny = y + dy[d]
        if nx < 0 or nx >= n or ny < 0 or ny >= n or triangle[ny][nx] != 0 :
            d = (d+1) % 3
        x = x + dx[d]
        y = y + dy[d]
    
    
    # 2중 for문 돌면서 새로운 1차원 리스트에 순서 저장 
    for i in range(n):
        for j in range(i+1):
            answer.append(triangle[i][j])
    
    return answer